R/print.R

```diff
@@ -281,10 +281,207 @@ print.bbi_nonmem_summary <- function(x, .digits = 3, .fixed = FALSE, .off_diag =
   if (!is.null(.nrow)) cat_line(glue("... {orig_rows - .nrow} more rows"), col = "grey")
 }
 
+
+#' @describeIn print_bbi Prints a high level summary of a model from a `bbi_nmboot_summary` object
+#' @importFrom purrr map_chr
+#' @importFrom cli cat_line
+#' @importFrom dplyr mutate_if
+#' @importFrom checkmate assert_number
+#'
+#' @param .digits Number of significant digits to use for parameter table. Defaults to 3.
+#' @param .nrow If `NULL`, the default, print all rows of the parameter table.
+#'   Otherwise, prints only `.nrow` rows.
+#' @export
+print.bbi_nmboot_summary <- function(x, .digits = 3, .nrow = 10, ...) {
+
+  # print top line info
+  .d <- x[[SUMMARY_DETAILS]]
+  cli_h1("Based on")
+  cat_line(paste("Model:", col_blue(x$based_on_model_path)))
+  cat_line(paste("Dataset:", col_blue(x$based_on_data_set)))
+
+  cli_h1("Run Specifications")
+  strat_cols <- if(rlang::is_empty(x$strat_cols)){
+    "None"
+  }else{
+    paste(x$strat_cols, collapse = ", ")
+  }
+  names(strat_cols) <- "Stratification Columns"
+
+  seed <- c("Seed" = x$seed)
+  n_samples <- c("Number of model runs" = x$n_samples)
+
+  iwalk(c(n_samples, strat_cols, seed),
+        ~ cat_bullet(paste0(.y, ": ", col_blue(.x))))
+
+  cli_h1("Run Summary")
+
+  # TODO: confirm this is appropriate for only_sim (unsure where this comes from)
+  only_sim <- isTRUE("only_sim" %in% names(.d))
+  if (only_sim) {
+    cat_line("No Estimation Methods (ONLYSIM)\n")
+  } else {
+    cli::cat_line("Estimation Method(s):\n")
+    purrr::walk(paste(x$estimation_method, "\n"), cli::cat_bullet, bullet = "en_dash")
+  }
+
+  # check heuristics
+  .h <- x[[SUMMARY_HEURISTICS]]
+  heuristics_cols <- names(.h)[!grepl(ABS_MOD_PATH, names(.h))]
+  heuristics <- purrr::map_dfr(heuristics_cols, function(col){
+    tibble(heuristic = col, any_found = any(.h[[col]]), n_found = sum(.h[[col]]))
+  })
+
+  if (any(heuristics$any_found)) {
+    cli::cat_line("**Heuristic Problem(s) Detected:**\n", col = "red")
+    heuristics_found <- heuristics$heuristic[which(heuristics$any_found)]
+    heuristics_n <- heuristics$n_found[which(heuristics$any_found)]
+    heuristics_perc <- (heuristics_n/n_samples) * 100
```

Comment:
I'd like to see the numbers rather than percents in the heuristic summary.  Also, I'd reframe this as a summary of the bootstrap runs rather than *alert* problems detected. 

Current output:
```r
**Heuristic Problem(s) Detected:**

– parameter_near_boundary (0.2%)
– hessian_reset (1.3%)
– has_final_zero_gradient (0.2%)
– prderr (0.1%)
```

Proposed: 

```r
Bootstrap run summary: 

- Number of runs: 1000
– parameter_near_boundary: 2 (0.2%)
– hessian_reset: 13 (1.3%)
– has_final_zero_gradient: 2 (0.2%)
– prderr: 1 (0.1%)

```

Reply:
Id like @seth127's approval here, just because a potential consistency concern (might want to remove both or neither)

Reply:
I would defer to @kylebaron on this display. I think these bootstraps are different enough from a regular model that a little inconsistency between the print output isn't a problem.

Reply:
@kylebaron let me know what you think of this structure. I moved the "number of runs" to the run summary instead of the specification section like you suggested. It does seem like a specification, but I could see it being more useful/digestible next to the heuristics. I changed the highlighting to just the values too since they are seemingly more of an expectation (to a degree) rather than a warning.

![image](https://github.com/metrumresearchgroup/bbr/assets/18685116/b86c6a17-6705-48ce-a73a-f45afd2f8c53)

Reply:
This looks great, @barrettk. Very clean summary of what was going on in the run. This is great for a first release. 

One cosmetic tweak that I might make is add some white space padding around `hessian_reset:` to get the summary numbers to line up; and might add some padding around the counts in the summary numbers. That's optional cosmetics but shouldn't be too hard to do. I think overall looks great. 

I guess one other small thing: the words `Run Summary` show up in two places pretty close together. Maybe change the second one to `Run Status` or just `Status`?

Thanks for making all the changes. 

<img width="669" alt="Screenshot 2024-05-01 at 14 04 41" src="https://github.com/metrumresearchgroup/bbr/assets/16908301/80001fd1-9860-438d-a6ff-6e1f724c40b6">